### PR TITLE
Bubbling up decryption errors

### PR DESCRIPTION
**Pull request by bennyn**

### CODE REVIEW

File: test/session/SessionSpec.coffee [hunk 1]
```diff
@@ -555,7 +555,7 @@ describe 'Session', ->
       bob = s
 
       Promise.all(alices.map (a) ->
-        return new Promise (resolve, reject) ->
+        return new Promise (resolve) ->
           Promise.all([0..900].map(-> a.encrypt 'hello'))
```
Comment (lipis) [inline test/session/SessionSpec.coffee]:
```coffeescript
Promise.all [0..900].map -> a.encrypt 'hello'
```

File: test/session/SessionSpec.coffee [hunk 2]
```diff
@@ -73,7 +73,7 @@ describe 'Session', ->
       assert(alice.session_states[alice.session_tag].state.recv_chains.length is 1)
       assert_serialise_deserialise alice_ident, alice
 
-  it 'encrypts and decrypts messsages', (done) ->
+  it 'encrypts and decrypts messages', (done) ->
     [alice_ident, bob_ident] = [0..1].map(-> Proteus.keys.IdentityKeyPair.new())
     [alice_store, bob_store] = [0..1].map(-> new TestStore Proteus.keys.PreKey.generate_prekeys 0, 10)
```
Comment (lipis) [inline test/session/SessionSpec.coffee]: parenthesis could go here as well..

File: src/proteus/session/Session.coffee [hunk 3]
```diff
@@ -193,30 +193,27 @@ module.exports = class Session
...
-            throw decrypt_error
+          else
+            return resolve @_decrypt_prekey_message envelope, msg, prekey_store
+
+  _decrypt_prekey_message: (envelope, msg, prekey_store) =>
+    return Promise.resolve()
+    .then =>
+      return @_decrypt_cipher_message envelope, msg.message
+    .catch (error) =>
+      if (error instanceof DecryptError.InvalidSignature or error instanceof DecryptError.InvalidMessage)
```
Comment (lipis) [inline src/proteus/session/Session.coffee]: `()`

File: src/proteus/session/Session.coffee [hunk 4]
```diff
@@ -189,34 +188,30 @@ module.exports = class Session
...
-
-            @_insert_session_state msg.message.session_tag, state
-            @pending_prekey = null
-
-            return resolve plaintext
-          .catch =>
-            throw decrypt_error
+          throw new DecryptError.RemoteIdentityChanged if msg.identity_key.fingerprint() isnt @remote_identity.fingerprint()
+          return resolve @_decrypt_prekey_message envelope, msg, prekey_store
+        else
```
Comment (herrmannplatz) [inline src/proteus/session/Session.coffee]: We should get rid of this `new Promise` pattern since its error-prone. see https://jsfiddle.net/ybmLrzko/

Comment (bennyn) [inline src/proteus/session/Session.coffee]: I was just about to ask you on this! 👍

**Pull request merged**